sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteWithExpression.scala

```diff
@@ -35,56 +35,82 @@ object RewriteWithExpression extends Rule[LogicalPlan] {
   override def apply(plan: LogicalPlan): LogicalPlan = {
     plan.transformWithPruning(_.containsPattern(WITH_EXPRESSION)) {
       case p if p.expressions.exists(_.containsPattern(WITH_EXPRESSION)) =>
-        var newChildren = p.children
-        var newPlan: LogicalPlan = p.transformExpressionsUp {
-          case With(child, defs) =>
-            val refToExpr = mutable.HashMap.empty[Long, Expression]
-            val childProjections = Array.fill(newChildren.size)(mutable.ArrayBuffer.empty[Alias])
+        val inputPlans = p.children.toArray
+        var newPlan: LogicalPlan = p.mapExpressions { expr =>
+          rewriteWithExprAndInputPlans(expr, inputPlans)
+        }
+        newPlan = newPlan.withNewChildren(inputPlans.toIndexedSeq)
+        if (p.output == newPlan.output) {
+          newPlan
+        } else {
+          Project(p.output, newPlan)
+        }
+    }
+  }
 
-            defs.zipWithIndex.foreach { case (CommonExpressionDef(child, id), index) =>
-              if (CollapseProject.isCheap(child)) {
-                refToExpr(id) = child
-              } else {
-                val childProjectionIndex = newChildren.indexWhere(
-                  c => child.references.subsetOf(c.outputSet)
-                )
-                if (childProjectionIndex == -1) {
-                  // When we cannot rewrite the common expressions, force to inline them so that the
-                  // query can still run. This can happen if the join condition contains `With` and
-                  // the common expression references columns from both join sides.
-                  // TODO: things can go wrong if the common expression is nondeterministic. We
-                  //       don't fix it for now to match the old buggy behavior when certain
-                  //       `RuntimeReplaceable` did not use the `With` expression.
-                  // TODO: we should calculate the ref count and also inline the common expression
-                  //       if it's ref count is 1.
-                  refToExpr(id) = child
-                } else {
-                  val alias = Alias(child, s"_common_expr_$index")()
-                  childProjections(childProjectionIndex) += alias
-                  refToExpr(id) = alias.toAttribute
-                }
-              }
-            }
+  private def rewriteWithExprAndInputPlans(
+      e: Expression,
+      inputPlans: Array[LogicalPlan]): Expression = {
+    if (!e.containsPattern(WITH_EXPRESSION)) return e
+    e match {
+      case w: With =>
+        // Rewrite nested With expression in CommonExpressionDef first.
+        val defs = w.defs.map(rewriteWithExprAndInputPlans(_, inputPlans))
```

Comment:
Now that we have "manual" recursion (instead of `transformExpressionsUp()`), shall we deal with nested `With`s in `w.child` too?

Reply:
Actually, the current logic seems to behave correctly if there is an inner `With` in an outer `With`'s `child` and the inner has a definition with a reference to an outer definition . (The previous `transformExpressionsUp()` had issues in that case.) But the rule is not idempotent, so maybe we should recurse into `w.child` after replacing `CommonExpressionRef`s?

Reply:
This is a good catch! It seems doesn't matter when to recurse into `w.child`, either before replacing `CommonExpressionRef` or after is fine?

Reply:
maybe before is better, as the expression tree may be much larger after replacing  `CommonExpressionRef`

Reply:
I'm not sure. E.g. if we have `With(With(x + x, Seq(x = y + y)), Seq(y = a + 1))` where `x` and `y` are references and `a` is an attribute and we would recurse into `With(x + x, Seq(x = y + y))` before replacing the `y` references to actual attributes that aliases `a + 1` then the `childProjectionIndex` won't find the right child, will it? But an UT covering this case would be good. :)

Reply:
oh correlated nested `With`! I'm not sure if we want to support it or not... But at least we should fail if we don't want to support it.